packages/metadata-hash/src/metadata-hash.ts

```diff
@@ -1,367 +1,111 @@
-      return [
-        {
-          tag: "bitSequence",
-          value: { numBytes, leastSignificantBitFirst },
-        },
-      ]
+  const getTypeRef = (frameId: number): TypeRef => {
+    const { def } = definitions.get(frameId)!
+
+    if (def.tag === "primitive") return { tag: def.value.tag, value: undefined }
+
+    if (def.tag === "compact") {
+      const primitive = getPrimitive(def.value)
+      const tag = compactTypeRefs[primitive as "u8"]
+      if (!tag) throw new Error("Invalid primitive for Compact")
+      return { tag, value: undefined }
     }
-    case "primitive":
-    case "compact":
-      throw new Error("Primitives and Compacts should be filtered out")
-  }
-}
 
-export const buildMetadataHash = (
-  metadata: V15,
-  info: ExtraInfo,
-): Uint8Array => {
-  const keyHash = h64(mergeUint8(v15.enc(metadata), extraInfo.enc(info)))
-  if (cache.has(keyHash)) {
-    return cache.get(keyHash)!
+    return accessibleTypes.has(frameId)
+      ? { tag: "perId", value: accessibleTypes.get(frameId)! }
+      : { tag: "void", value: undefined }
   }
 
-  const accessibleTypes: number[] = []
-  pruneAndCollectAccessibleTypes(
-    metadata,
-    accessibleTypes,
-    metadata.extrinsic.call,
-  )
-  pruneAndCollectAccessibleTypes(
-    metadata,
+  const trimmedTree = getTypeTree(
+    definitions,
     accessibleTypes,
-    metadata.extrinsic.address,
-  )
-  pruneAndCollectAccessibleTypes(
-    metadata,
-    accessibleTypes,
-    metadata.extrinsic.signature,
-  )
-  metadata.extrinsic.signedExtensions.forEach((se) => {
-    pruneAndCollectAccessibleTypes(metadata, accessibleTypes, se.type)
-    pruneAndCollectAccessibleTypes(
-      metadata,
-      accessibleTypes,
-      se.additionalSigned,
-    )
-  })
-  accessibleTypes.sort((a, b) => a - b)
-
-  const typeTree: Lookup = []
-  accessibleTypes.forEach((ty, idx) => {
-    const frameType = metadata.lookup.find((x) => x.id === ty)!
-    constructTypeDef(metadata, accessibleTypes, ty).forEach((tyDef) => {
-      typeTree.push({
-        path: frameType.path,
-        typeId: idx,
-        typeDef: tyDef,
-      })
-    })
-  })
+    getTypeRef,
+    getPrimitive,
+  ).map(lookupType.enc)
 
-  typeTree.sort((a, b) => {
-    if (a.typeId !== b.typeId) return a.typeId - b.typeId // in general
-    // should only happen for variants
-    if (a.typeDef.tag !== "enumeration" || b.typeDef.tag !== "enumeration")
-      throw new Error("Found two types with same id")
-    return a.typeDef.value.index - b.typeDef.value.index
-  })
+  const cachedKey = h64(mergeUint8([...trimmedTree, extraInfo.enc(info)]))
+  const cached = cache.get(cachedKey)
+  if (cached) return cached
 
   // let's build the lookup hash
-  const nodes = typeTree.map((x) => Blake3256(lookupType.enc(x)))
-  while (nodes.length > 1) {
-    const right = nodes.pop()!
-    const left = nodes.pop()!
-    nodes.unshift(Blake3256(mergeUint8(left, right)))
+  const hashesTree = new Array(trimmedTree.length * 2 - 1)
```

Comment:
If `trimmedTree.length === 0` this will go nuts. Maybe we could directly return the 0's hash here to prevent issues (it is not expected to happen though, but let's be picky).